### PR TITLE
improve cache performance for single date queries

### PR DESCRIPTION
**Pull request by mzruya**
The Holidays gem performs a lot of single date queries, these are done by using the date extensions and calling `Date#holiday?` or by calling `Holidays.on?`.

The single date queries delegate to `Holidays.between(date, date)`, which in turn uses the underlying cache (if available).

The majority of use cases I have seen for the holiday gem in my organization were single date queries, which are very inefficient with the given cache structure.

The cache is currently an array of holidays, its lookup method has a `O(n)` time complexity where `n` is the amount of holidays currently in the cache.
As one would expect the performance impact is linear to the size of the cache.

This simple optimization changes the cache's data store from an array to a lookup table.

## Benchmarks
```ruby
date = Date.new(2017, 1, 1)
Holidays.cache_between(
  Date.new(2017, 1, 1), 
  Date.new(2017 + years , 1, 1), 
  :us, 
  :observed
)
Benchmark.measure { 10000.times.each { date.holiday?(:us, :observed) } }.real
```

As a baseline, this computation takes **4.280** seconds without any cache.

### Before

| Years  | Cache Size | Time (seconds) |
| ------ | ----------- | ---------------- |
| 1 | 11 | 0.069 |
| 5 | 51 | 0.168 |
| 10 | 101 | 0.292 |
| 15 | 151 |  0.437 |
| 20 | 201 | 0.626 ||
| 30 | 301 | 0.812 |
| 40 | 401 | 1.253 |
| 50 | 501 |  1.489 |
| 100 | 101 | 3.166 |
| 150 | 1501 | 4.403 |
| 200 | 2001 | 6.384 |

at a size of ~ 1500 the cache becomes detrimental for this use case and degrades performance.

### After

| Years  | Cache Size | Time (seconds) |
| ------ | ----------- | ---------------- |
| 1 | 11 | 0.041 |
| 5 | 51 | 0.035 |
| 10 | 101 | 0.031 |
| 15 | 151 | 0.032 |
| 20 | 201 | 0.032 |
| 30 | 301 | 0.035 |
| 40 | 401 | 0.032 |
| 50 | 501 | 0.032 |
| 100 | 1001 | 0.045 |
| 150 | 1501 | 0.042 |
| 200 | 2001 | 0.041 |

### CODE REVIEW

File: test/holidays/definition/repository/test_cache.rb [hunk 1]
```diff
@@ -7,20 +7,55 @@ def setup
     @subject = Holidays::Definition::Repository::Cache.new
   end
 
+  def test_find_supports_overlapping_holidays
```
Comment (mzruya) [inline test/holidays/definition/repository/test_cache.rb]: I'm not sure if this test case is valuable in the real world, but felt obliged due to the semantics of the data structures.

**Comment by ptrimble** [comment]
Wow, thank you for this! And thank you for your patience, I am returning from paternity leave so it's taking me a while to get back up to running speed.

The changes initially look great! I would like to finish up some in-progress work I left from before I took my time off and then this is next on my list. I'll be in touch soon!

**Comment by mzruya** [comment]
haha, well, you deserve a Mazel Tov! take your time :-)

**Comment by ttwo32** [comment]
Hi @mzruya !!
Firstly, I apologize for the delay in replying to you.
Thank you for this PR.

File: lib/holidays/definition/repository/cache.rb [hunk 2]
```diff
@@ -10,23 +10,36 @@ def cache_between(start_date, end_date, cache_data, options)
           raise ArgumentError unless cache_data
 
           @cache_range[options] = start_date..end_date
-          @cache[options] = cache_data
+          @cache[options] = cache_data.group_by { |holiday| holiday[:date] }
```
Comment (ttwo32) [inline lib/holidays/definition/repository/cache.rb]: It is a good idea to make cache_data Hash table. 👍

**Comment by ttwo32** [comment]
@mzruya LGTM. 
Please wait a response from @ptrimble . :)

File: test/holidays/definition/repository/test_cache.rb (hunk 1 above)
Comment (ppeble) [inline test/holidays/definition/repository/test_cache.rb]: 👍 

Good plan, I support this.

File: lib/holidays/definition/repository/cache.rb (hunk 2 above)
Comment (ppeble) [inline lib/holidays/definition/repository/cache.rb]: Agreed, great idea.

**Comment by ppeble** [comment]
Thank you again! Merged. 😃

**Pull request merged**